src/python/pants/option/config.py

```diff
@@ -230,10 +244,173 @@ def get_value(self, section: str, option: str) -> Optional[str]:
+    def format_str(value: str) -> str:
+      # Because dictionaries use the symbols `{}`, we must proactively escape the symbols so that
+      # .format() does not try to improperly interpolate.
+      escaped_str = value.replace("{", "{{").replace("}", "}}")
+      new_style_format_str = re.sub(
+        pattern=r"%\((?P<interpolated>[a-zA-Z_0-9]*)\)s",
+        repl=r"{\g<interpolated>}",
+        string=escaped_str,
+      )
+      return new_style_format_str.format(**self.defaults)
+
+    def recursively_format_str(value: str) -> str:
+      # It's possible to interpolate with a value that itself has an interpolation. We must fully
+      # evaluate all expressions for parity with configparser.
+      formatted_str = format_str(value)
+      if not re.search(r"%\([a-zA-Z_0-9]*\)s", formatted_str):
+        return formatted_str
+      return recursively_format_str(formatted_str)
+
+    return recursively_format_str(raw_value)
+
+  def _stringify_val(
+    self, raw_value: _TomlValue, *, interpolate: bool = True, list_prefix: Optional[str] = None,
+  ) -> str:
+    """For parity with configparser, we convert all values back to strings, which allows us to
+    avoid upstream changes to files like parser.py.
+
+    This is clunky. If we drop INI support, we should remove this and use native values."""
+    if isinstance(raw_value, str):
+      return self._possibly_interpolate_value(raw_value) if interpolate else raw_value
+
+    if isinstance(raw_value, list):
+      def stringify_list_member(member: _TomlPrimitve) -> str:
+        if not isinstance(member, str):
+          return str(member)
+        interpolated_member = self._possibly_interpolate_value(member) if interpolate else member
+        return f'"{interpolated_member}"'
+
+      list_members = ", ".join(stringify_list_member(member) for member in raw_value)
+      return f"{list_prefix or ''}[{list_members}]"
+
+    return str(raw_value)
+
+  @property
+  def sections(self) -> List[str]:
+    sections: List[str] = []
+
+    def recurse(mapping: Dict, *, parent_section: Optional[str] = None) -> None:
+      for section, section_values in mapping.items():
+        if not isinstance(section_values, dict):
+          continue
+        # We filter out "DEFAULT" and also check for the special `my_list_option.append` and
+        # `my_list_option.filter` syntax.
+        if section == "DEFAULT" or "append" in section_values or "filter" in section_values:
+          continue
+        section_name = section if not parent_section else f"{parent_section}.{section}"
+        if self._section_explicitly_defined(section_values):
+          sections.append(section_name)
+        recurse(section_values, parent_section=section_name)
+
+    recurse(self.values)
+    return sections
+
+  def has_section(self, section: str) -> bool:
+    return self._find_section(section) is not None
+
+  def has_option(self, section: str, option: str) -> bool:
+    try:
+      self.get_value(section, option)
```

Comment:
Technically, `has_option` is doing more work than necessary by relying on the `get_value()` implementation, e.g. it will interpolate values and `_stringify` them. But, I think that's fine to make this method much simpler. Otherwise, it has extremely high duplication of `get_value()`.